numpy/_core/src/multiarray/datetime.c

```diff
@@ -2818,85 +2819,247 @@ convert_datetime_to_pyobject(npy_datetime dt, PyArray_DatetimeMetaData *meta)
 }
 
 /*
- * Converts a timedelta into a PyObject *.
+ * We require that if d is a PyDateTime, then
+ * hash(numpy.datetime64(d)) == hash(d).
+ * Where possible, convert dt to a PyDateTime and hash it.
  *
- * Not-a-time is returned as the string "NaT".
- * For microseconds or coarser, returns a datetime.timedelta.
- * For units finer than microseconds, returns an integer.
+ * NOTE: "equals" across PyDate, PyDateTime and np.datetime64 is not transitive:
+ * datetime.datetime(1970, 1, 1) == np.datetime64(0, 'us')
+ * np.datetime64(0, 'us') == np.datetime64(0, 'D')
+ * datetime.datetime(1970, 1, 1) != np.datetime64(0, 'D')  # date, not datetime!
+ *
+ * But:
+ * datetime.date(1970, 1, 1) == np.datetime64(0, 'D')
+ *
+ * For hash(datetime64(0, 'D')) we could return either PyDate.hash or PyDateTime.hash.
+ * We choose PyDateTime.hash to match datetime64(0, 'us')
  */
-NPY_NO_EXPORT PyObject *
-convert_timedelta_to_pyobject(npy_timedelta td, PyArray_DatetimeMetaData *meta)
+NPY_NO_EXPORT npy_hash_t
+datetime_hash(PyArray_DatetimeMetaData *meta, npy_datetime dt)
 {
-    npy_timedelta value;
-    int days = 0, seconds = 0, useconds = 0;
+    npy_hash_t res, obj_hash;
+    npy_datetimestruct dts;
 
-    /*
-     * Convert NaT (not-a-time) into None.
-     */
-    if (td == NPY_DATETIME_NAT) {
-        Py_RETURN_NONE;
+    if (dt == NPY_DATETIME_NAT || meta->base == NPY_FR_GENERIC) {
```

Comment:
This looks wrong.  For `generic` unit you need to return the value as if an int64.

Reply:
I'm happy to do that, but is it even possible to construct a datetime64 with base == generic?

Reply:
Hmmm, good point, maybe a comment?  The timedelta error is also interesting, but I am happy to try it.  Users shouldn't explicitly create them (if that is even easy), but rather implicitly/transiently when doing math with integers.
(And even there, we may stop using that path eventually.)

Reply:
I went with your suggestion anyway :-)